From 7384d7cbaaeaae7056dcd250ec8472d8ebe8599c Mon Sep 17 00:00:00 2001
From: Duy Bao Dang <duy.dang.yb@renesas.com>
Date: Thu, 7 Jan 2021 14:55:35 +0700
Subject: [PATCH 215/433] drm: adv7511: Workaround for I2C address conflict
 issue

This patch uses an alternative value for CEC Memory address to avoid
I2C address conflict between ADV7511 and OV5640 CMOS camera that may
happen on the iWave RZ/G1{M,N} boards.

With the modification, I2C address of CEC Memory will be changed from
0x3c to 0x3d.

Signed-off-by: Duy Bao Dang <duy.dang.yb@renesas.com>
Signed-off-by: Hao Bui <hao.bui.yg@renesas.com>
---
 drivers/gpu/drm/i2c/adv7511.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i2c/adv7511.c b/drivers/gpu/drm/i2c/adv7511.c
index 4300e27..52feb74 100644
--- a/drivers/gpu/drm/i2c/adv7511.c
+++ b/drivers/gpu/drm/i2c/adv7511.c
@@ -12,6 +12,7 @@
 #include <linux/module.h>
 #include <linux/regmap.h>
 #include <linux/slab.h>
+#include <linux/of.h>
 
 #include <drm/drmP.h>
 #include <drm/drm_crtc_helper.h>
@@ -54,6 +55,8 @@ struct adv7511 {
 static const int edid_i2c_addr = 0x7e;
 static const int packet_i2c_addr = 0x70;
 static const int cec_i2c_addr = 0x78;
+/* alternative address to avoid conflict */
+static const int cec_i2c_addr_alt = 0x7a;
 
 static struct adv7511 *encoder_to_adv7511(struct drm_encoder *encoder)
 {
@@ -952,7 +955,18 @@ static int adv7511_probe(struct i2c_client *i2c, const struct i2c_device_id *id)
 	regmap_write(adv7511->regmap, ADV7511_REG_EDID_I2C_ADDR, edid_i2c_addr);
 	regmap_write(adv7511->regmap, ADV7511_REG_PACKET_I2C_ADDR,
 		     packet_i2c_addr);
-	regmap_write(adv7511->regmap, ADV7511_REG_CEC_I2C_ADDR, cec_i2c_addr);
+	/*
+	 * use alternative address for CEC Memory address in case of
+	 * RZ/G1{M,N} boards to avoid possible conflict to OV5640 address
+	 */
+	if (of_machine_is_compatible("renesas,r8a7743") ||
+		    of_machine_is_compatible("renesas,r8a7744"))
+		regmap_write(adv7511->regmap, ADV7511_REG_CEC_I2C_ADDR,
+				cec_i2c_addr_alt);
+	else
+		regmap_write(adv7511->regmap, ADV7511_REG_CEC_I2C_ADDR,
+				cec_i2c_addr);
+
 	adv7511_packet_disable(adv7511, 0xffff);
 
 	adv7511->i2c_main = i2c;
-- 
2.7.4

